[DED] Add transform to the Data Exfiltration Detection package

sodhikirti07 opened the pull request:
<!-- Type of change
Please label this PR with one of the following labels, depending on the scope of your change:
- Bug
- Enhancement
- Breaking change
- Deprecation
-->

## Proposed commit message

<!-- Mandatory
Explain here the changes you made on the PR.

Please explain:

- WHAT: patterns used, algorithms implemented, design architecture, message processing, etc.
- WHY:  the rationale/motivation for the changes

This text will be pasted into the squash dialog when the change is committed and will be
a long term historical record of the change to help future contributors understand the
change, please help them by making it clear and comprehensive, they may be you.

If the commit title is adequate to describe both of these things, The text here may be omitted
or replaced with "See title". The title of the PR will be used as the commit message title when
the merge is made and the "See title" marker will be removed if present.

The text here and the PR title will be subject to the PR review process.
-->
Added pivot transform to the DED package.
## Checklist

- [ ] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [ ] I have verified that all data streams collect metrics or logs.
- [ ] I have added an entry to my package's `changelog.yml` file.
- [ ] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).

## Author's Checklist

<!-- Recommended
Add a checklist of things that are required to be reviewed in order to have the PR approved
-->
- Added a Pivot transform to handle duplicates in unidirectional network logs, which may occur when users have multiple integrations enabled. Tested the transform using `elastic-package`, and it is functioning correctly (see screenshot below in comments).
- Updated the `README.md` with setup instructions for the transform. Also included instructions on customizing filters in the transform search query to manage unwanted noise.
- Updated `manifest.yml` and `changelog.yml`.

## How to test this PR locally

<!-- Recommended
Explain here how this PR will be tested by the reviewer: commands, dependencies, steps, etc.
-->

## Related issues

<!-- Recommended
Link related issues below. Insert the issue link or reference after the word "Closes" if merging this should automatically close it.

- Closes #123
- Relates #123
- Requires #123
- Supersedes #123
-->
- 

## Screenshots

<!-- Optional
Add here screenshots presenting:
- Kibana UI forms presenting configuration options exposed by the integration
- dashboards with collected metrics or logs
-->

sodhikirti07 commented:
### Screenshots:

**Package Overview:**
<img width="1489" alt="image" src="https://github.com/user-attachments/assets/b50b82b8-6043-49c7-bbfa-38c5e167b83b">


**DED transform:**
![image](https://github.com/user-attachments/assets/b81d029b-62da-4e5d-b6ac-228657fa3c97)

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/integrations/builds/14438)
* Commit: 762b55ddb1e04ffeebe5ff4085078cec294f9b8f

<!--buildkite-pr-comment-integrations
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/14438","number":14438,"commit":"762b55ddb1e04ffeebe5ff4085078cec294f9b8f"}],"number":14438}
buildkite-pr-comment-->

jmcarlock commented:
@sodhikirti07 Is the purpose of changing the data feeds to prepare for [bi-directional connection analysis](https://github.com/elastic/security-team/issues/8152#issuecomment-2266016953)? Will this affect the number of false positives for the currently existing anomaly detectors?

sodhikirti07 commented:
> Is the purpose of changing the data feeds to prepare for https://github.com/elastic/security-team/issues/8152#issuecomment-2266016953? 

@jmcarlock Our package currently doesn't support bi-directional network flows; we plan to add this as a new feature in the future (I've added this statement in the README as well). The search query of the transform contains all the necessary filters, eliminating the need for them at the datafeed level. This will also make it easier for users to customize the search query. 


> Will this affect the number of false positives for the currently existing anomaly detectors?

I've added a new filter for `process.name` to exclude noise from Elastic native tools, which should help reduce some noise. Additionally, users can customize the transform by adding or removing filters to meet their individual requirements.

qn895 commented:
Transform yml files LGTM 🎉

susan-shu-c commented:
Just took a look, this is great!
Question, what version of Kibana were you developing on? Since the manifest file says this support [^Kibana 8.9.0](https://github.com/elastic/integrations/pull/10726/files#diff-ea8dcf4d2fbb7295f83facd18e6c0a2b0db87d8cf70ec2547ddc5d8f155f7b19R15), it will be good if we can confirm this update will run as expected

sodhikirti07 commented:
> Question, what version of Kibana were you developing on? Since the manifest file says this support [^Kibana 8.9.0](https://github.com/elastic/integrations/pull/10726/files#diff-ea8dcf4d2fbb7295f83facd18e6c0a2b0db87d8cf70ec2547ddc5d8f155f7b19R15), it will be good if we can confirm this update will run as expected

@susan-shu-c Good question and thanks for reminding me! I tested the package on `v8.11.0`, however, as this package has `alias` spec in the transform, it'll only work with `^v8.10.1`, similar to that of [beaconing](https://github.com/elastic/integrations/blob/920aa511332e242e0aae5cd0b3d9794badc5571d/packages/beaconing/manifest.yml#L14). Going to test and make this change soon.

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/integrations/builds/14493)
* Commit: 637cb6e9d2fac428db6ba8b8e0e7e917aefba791

### History
* :green_heart: [Build #14438](https://buildkite.com/elastic/integrations/builds/14438) succeeded 762b55ddb1e04ffeebe5ff4085078cec294f9b8f

<!--buildkite-pr-comment-integrations
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/14493","number":14493,"commit":"637cb6e9d2fac428db6ba8b8e0e7e917aefba791"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/14438","number":14438,"commit":"762b55ddb1e04ffeebe5ff4085078cec294f9b8f"}],"number":14493}
buildkite-pr-comment-->

jmcarlock commented on packages/ded/elasticsearch/transform/pivot_transform/transform.yml:
```diff
@@ -0,0 +1,92 @@
+        - terms:
+            'network.direction': [ "internal","inbound" ]
+        - terms:
+            'process.name': [ "elastic-agent.exe","elastic-agent","metricbeat.exe","metricbeat","filebeat.exe","filebeat","packetbeat.exe","packetbeat" ]
```
Should you add "winlogbeat.exe" and "winlogbeat" here? This is from my test on Elastic Cloud.

<img width="1375" alt="Screenshot 2024-08-08 at 2 41 15 PM" src="https://github.com/user-attachments/assets/74f6903a-2e09-478e-9298-21872485f237">

sodhikirti07 replied:
Done! Thank you for pointing this out!

susan-shu-c replied:
Good catch!

jmcarlock commented:
This works great, @sodhikirti07 !  I just have two other small suggestions:
- Move the "Install assets" installation steps after "Add the Integration Package", as I had to do this before using the transform.
  <img width="1304" alt="Screenshot 2024-08-08 at 2 23 57 PM" src="https://github.com/user-attachments/assets/7c1cf5bf-43c1-4f11-bcd3-09b54c9b859c">
- I would suggest adding a header to "To customize the filters in the Data Exfiltration transform..." and referencing it in the instructions in step 3. Otherwise users might not see it.

sodhikirti07 commented:
> Move the "Install assets" installation steps after "Add the Integration Package", as I had to do this before using the transform.

@jmcarlock Which version of Kibana did you test this on? Was it a local setup, as I didn't have to install any assets separately?

jmcarlock commented:
> > Move the "Install assets" installation steps after "Add the Integration Package", as I had to do this before using the transform.
> 
> @jmcarlock Which version of Kibana did you test this on? Was it a local setup, as I didn't have to install any assets separately?

Elastic Cloud v8.14.3

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/integrations/builds/14495)
* Commit: b0642416dbf7aaeb3f69193e7d5d083d2c52680c

### History
* :green_heart: [Build #14493](https://buildkite.com/elastic/integrations/builds/14493) succeeded 637cb6e9d2fac428db6ba8b8e0e7e917aefba791
* :green_heart: [Build #14438](https://buildkite.com/elastic/integrations/builds/14438) succeeded 762b55ddb1e04ffeebe5ff4085078cec294f9b8f

<!--buildkite-pr-comment-integrations
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/14495","number":14495,"commit":"b0642416dbf7aaeb3f69193e7d5d083d2c52680c"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/14493","number":14493,"commit":"637cb6e9d2fac428db6ba8b8e0e7e917aefba791"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/14438","number":14438,"commit":"762b55ddb1e04ffeebe5ff4085078cec294f9b8f"}],"number":14495}
buildkite-pr-comment-->

sodhikirti07 commented:
> Elastic Cloud v8.14.3. It's possible that it showed up because I uploaded it as a zip.

I added the "Install assets" step as you suggested, however, I didn't have to follow this to enable the transform. I think it's still best to keep it after the step 2.

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/integrations/builds/14496)
* Commit: af7db597d754c92db02133889fcd604fbba4c93c

### History
* :green_heart: [Build #14495](https://buildkite.com/elastic/integrations/builds/14495) succeeded b0642416dbf7aaeb3f69193e7d5d083d2c52680c
* :green_heart: [Build #14493](https://buildkite.com/elastic/integrations/builds/14493) succeeded 637cb6e9d2fac428db6ba8b8e0e7e917aefba791
* :green_heart: [Build #14438](https://buildkite.com/elastic/integrations/builds/14438) succeeded 762b55ddb1e04ffeebe5ff4085078cec294f9b8f

<!--buildkite-pr-comment-integrations
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/14496","number":14496,"commit":"af7db597d754c92db02133889fcd604fbba4c93c"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/14495","number":14495,"commit":"b0642416dbf7aaeb3f69193e7d5d083d2c52680c"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/14493","number":14493,"commit":"637cb6e9d2fac428db6ba8b8e0e7e917aefba791"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/integrations/builds/14438","number":14438,"commit":"762b55ddb1e04ffeebe5ff4085078cec294f9b8f"}],"number":14496}
buildkite-pr-comment-->

susan-shu-c reviewed (approved):
Great discussion, and thanks for addressing comments!

jmcarlock commented:
Tested with the updated transform and winlogbeat no longer appears, thank you @sodhikirti07!

elasticmachine commented:
Package ded - 2.2.0 containing this change is available at https://epr.elastic.co/search?package=ded
